File: Lecture037 Recursion Day7/subsequences.cpp
void solve(vector<string>& ans, string str, string output, int i) {
    //base case
    if(i>=str.length()) {
        if(output.length()>0)
        ans.push_back(output);
        return ;
        }
    
    //exclude
    solve(ans, str, output, i+1);
    //include
    output.push_back(str[i]);
    solve(ans, str, output, i+1);
}

vector<string> subsequences(string str){
	
	vector<string> ans;
    string output = "";
    solve(ans,str,output,0);
    return ans;
	
}

File: Lecture037 Recursion Day7/Subsequences.java
package com.company;
import  java.util.*;
public class Subsequences {
    static void solve(String str,int index, ArrayList<String> ans,String output){
        if(index>=str.length()){
            ans.add(output);
            return;
        }
        solve(str,index+1,ans,output);
        output = output+str.charAt(index);
        solve(str,index+1,ans,output);

    }
    public static ArrayList<String> subsequences(String str) {
        // Write your code here
        ArrayList<String> ans =new ArrayList<>();
        String output ="";
        solve(str,0,ans,output);
        ans.remove(0);
        Collections.sort(ans);
        return ans;
    }
    public static void main(String[] args) {
        String str ="abc";

        System.out.println(subsequences(str));
    }
}
